internal/dag/policy.go

```diff
@@ -125,7 +126,43 @@ func headersPolicyRoute(policy *contour_api_v1.HeadersPolicy, allowHostRewrite b
 func escapeHeaderValue(value string) string {
 	// Envoy supports %-encoded variables, so literal %'s in the header's value must be escaped.  See:
 	// https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#custom-request-response-headers
-	return strings.Replace(value, "%", "%%", -1)
+	// Only allow a specific set of known good Envoy dynamic headers to pass through unescaped
+	if !strings.Contains(value, "%") {
```

Comment:
Should this check that it begins / ends with the `%`?

Reply:
i think the header value can be a combination of variables and static content e.g. `foo: %HOSTNAME% bar`

Reply:
This just avoids all of the processing below if there is no % in the header value.

There are tests to ensure that %'s that are not part of a dynamic header are properly escaped, including with multiple dynamic headers and other non-dynamic content.